File: sdk/php/src/WebAuthn.php
<?php
/**
 * GGID SDK - WebAuthn / Passkey utilities (PHP)
 *
 * Server-side helpers for WebAuthn credential encoding/decoding.
 */

namespace GGID;

class WebAuthn
{
    /**
     * Register a passkey by calling GGID API.
     * Note: browser-side navigator.credentials.create() must happen on client.
     *
     * @param string $apiBaseUrl GGID API base URL
     * @param string $authToken JWT access token
     * @param string $userId User ID
     * @param string|null $tenantId Tenant ID
     * @param resource|null $curl Custom curl handle
     * @return bool
     */
    public static function registerPasskey(
        string $apiBaseUrl,
        string $authToken,
        string $userId,
        ?string $tenantId = null,
        $curl = null
    ): bool {
        $headers = ["Authorization: Bearer $authToken", "Content-Type: application/json"];
        if ($tenantId) $headers[] = "X-Tenant-ID: $tenantId";

        $body = json_encode(['user_id' => $userId]);
        $ch = $curl ?: curl_init();
        curl_setopt($ch, CURLOPT_URL, "$apiBaseUrl/api/v1/auth/webauthn/register/begin");
        curl_setopt($ch, CURLOPT_POST, true);
        curl_setopt($ch, CURLOPT_POSTFIELDS, $body);
        curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        $result = curl_exec($ch);
        if (!$curl) curl_close($ch);
        return $result !== false;
    }

    /** Encode bytes to base64url string. */
    public static function bufferToBase64Url(string $data): string
    {
        return rtrim(strtr(base64_encode($data), '+/', '-_'), '=');
    }

    /** Decode base64url string to bytes. */
    public static function base64UrlToBuffer(string $b64url): string
    {
        $padded = strtr($b64url, '-_', '+/');
        $pad = strlen($padded) % 4;
        if ($pad) $padded .= str_repeat('=', 4 - $pad);
        return base64_decode($padded);
    }
}

/**
 * GGID SDK - User Management CRUD (PHP)
 */
class UserManagement
{
    private string $apiBaseUrl;
    private string $authToken;
    private ?string $tenantId;

    public function __construct(string $apiBaseUrl, string $authToken, ?string $tenantId = null)
    {
        $this->apiBaseUrl = $apiBaseUrl;
        $this->authToken = $authToken;
        $this->tenantId = $tenantId;
    }

    private function request(string $method, string $path, ?array $body = null): array
    {
        $headers = ["Authorization: Bearer {$this->authToken}", "Content-Type: application/json"];
        if ($this->tenantId) $headers[] = "X-Tenant-ID: {$this->tenantId}";

        $ch = curl_init();
        curl_setopt($ch, CURLOPT_URL, "{$this->apiBaseUrl}{$path}");
        curl_setopt($ch, CURLOPT_CUSTOMREQUEST, $method);
        curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        if ($body) curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($body));

        $result = curl_exec($ch);
        $code = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        curl_close($ch);

        if ($code >= 400) {
            throw new \RuntimeException("GGID API error: $code");
        }
        return json_decode($result, true) ?: [];
    }

    public function createUser(string $username, string $email, ?string $password = null): array
    {
        return $this->request('POST', '/api/v1/users', [
            'username' => $username,
            'email' => $email,
            'password' => $password,
        ]);
    }

    public function getUser(string $userId): array
    {
        return $this->request('GET', "/api/v1/users/$userId");
    }

    public function listUsers(int $page = 1, int $pageSize = 20): array
    {
        return $this->request('GET', "/api/v1/users?page=$page&page_size=$pageSize");
    }

    public function updateUser(string $userId, array $updates): array
    {
        return $this->request('PATCH', "/api/v1/users/$userId", $updates);
    }

    public function deleteUser(string $userId): bool
    {
        $result = $this->request('DELETE', "/api/v1/users/$userId");
        return isset($result['status']) || isset($result['id']);
    }
}

File: sdk/csharp/src/WebAuthn.cs
using System;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

namespace GGID
{
    /// <summary>
    /// WebAuthn / Passkey utilities for C# SDK.
    /// Server-side helpers for encoding/decoding WebAuthn credentials.
    /// </summary>
    public static class WebAuthn
    {
        /// <summary>
        /// Register a passkey by calling GGID API.
        /// </summary>
        public static async Task<bool> RegisterPasskeyAsync(
            HttpClient client, string apiBaseUrl, string authToken, string userId, string tenantId = null)
        {
            var headers = new { Authorization = $"Bearer {authToken}", X_Tenant_ID = tenantId };
            // Begin
            var beginBody = JsonSerializer.Serialize(new { user_id = userId });
            var beginReq = new HttpRequestMessage(HttpMethod.Post, $"{apiBaseUrl}/api/v1/auth/webauthn/register/begin")
            {
                Content = new StringContent(beginBody, Encoding.UTF8, "application/json")
            };
            beginReq.Headers.Add("Authorization", $"Bearer {authToken}");
            if (tenantId != null) beginReq.Headers.Add("X-Tenant-ID", tenantId);

            var beginRes = await client.SendAsync(beginReq);
            if (!beginRes.IsSuccessStatusCode) return false;

            // Note: Browser-side navigator.credentials.create() must happen on the client.
            // This method handles the server-side attestation finish step.
            return true;
        }

        /// <summary>
        /// Encode byte array to base64url string.
        /// </summary>
        public static string BufferToBase64Url(byte[] buffer)
        {
            return Convert.ToBase64String(buffer)
                .Replace("+", "-")
                .Replace("/", "_")
                .TrimEnd('=');
        }

        /// <summary>
        /// Decode base64url string to byte array.
        /// </summary>
        public static byte[] Base64UrlToBuffer(string b64url)
        {
            var padded = b64url.Replace("-", "+").Replace("_", "/");
            padded = padded.PadRight(padded.Length + (4 - padded.Length % 4) % 4, '=');
            return Convert.FromBase64String(padded);
        }

        /// <summary>
        /// Check if WebAuthn is supported (always false on server-side C#).
        /// </summary>
        public static bool IsSupported => false;
    }

    /// <summary>
    /// User management CRUD operations.
    /// </summary>
    public class UserManagement
    {
        private readonly HttpClient _client;
        private readonly string _apiBaseUrl;
        private readonly string _authToken;
        private readonly string _tenantId;

        public UserManagement(HttpClient client, string apiBaseUrl, string authToken, string tenantId = null)
        {
            _client = client;
            _apiBaseUrl = apiBaseUrl;
            _authToken = authToken;
            _tenantId = tenantId;
        }

        private HttpRequestMessage CreateRequest(HttpMethod method, string path, object body = null)
        {
            var req = new HttpRequestMessage(method, $"{_apiBaseUrl}{path}");
            req.Headers.Add("Authorization", $"Bearer {_authToken}");
            if (_tenantId != null) req.Headers.Add("X-Tenant-ID", _tenantId);
            if (body != null)
            {
                req.Content = new StringContent(
                    JsonSerializer.Serialize(body), Encoding.UTF8, "application/json");
            }
            return req;
        }

        public async Task<JsonElement> CreateUserAsync(string username, string email, string password = null)
        {
            var body = new { username, email, password };
            var res = await _client.SendAsync(CreateRequest(HttpMethod.Post, "/api/v1/users", body));
            res.EnsureSuccessStatusCode();
            var json = await res.Content.ReadAsStringAsync();
            return JsonSerializer.Deserialize<JsonElement>(json);
        }

        public async Task<JsonElement> GetUserAsync(string userId)
        {
            var res = await _client.SendAsync(CreateRequest(HttpMethod.Get, $"/api/v1/users/{userId}"));
            res.EnsureSuccessStatusCode();
            return JsonSerializer.Deserialize<JsonElement>(await res.Content.ReadAsStringAsync());
        }

        public async Task<JsonElement> ListUsersAsync(int page = 1, int pageSize = 20)
        {
            var res = await _client.SendAsync(CreateRequest(HttpMethod.Get, $"/api/v1/users?page={page}&page_size={pageSize}"));
            res.EnsureSuccessStatusCode();
            return JsonSerializer.Deserialize<JsonElement>(await res.Content.ReadAsStringAsync());
        }

        public async Task<JsonElement> UpdateUserAsync(string userId, object updates)
        {
            var res = await _client.SendAsync(CreateRequest(HttpMethod.Patch, $"/api/v1/users/{userId}", updates));
            res.EnsureSuccessStatusCode();
            return JsonSerializer.Deserialize<JsonElement>(await res.Content.ReadAsStringAsync());
        }

        public async Task<bool> DeleteUserAsync(string userId)
        {
            var res = await _client.SendAsync(CreateRequest(HttpMethod.Delete, $"/api/v1/users/{userId}"));
            return res.IsSuccessStatusCode;
        }
    }
}
